### PR TITLE
Migrate to new message processing logic with `core-processor`

### PR DESCRIPTION
**Pull request by breathx**
TODO:
- [x] write release notes
- [ ] write related issue
- [ ] open new issues based on research during this PR
- [ ] reset the network after merge

**Release notes**: This migrates to new logic of processing messages with functional approach and journalling. This allows us to use multiple ways to actually execute core logic on different environments (collator, validator and cumulus setups).

### CODE REVIEW

File: runner/src/runner.rs [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+            ext,
+            instrumented_code,
+            program.get_pages(),
+            &*memory,
+            settings.entry.into(),
+        );
+
+        // Parsing outcome.
+        let outcome = if let Err(e) = res {
+            let explanation = ext.error_explanation.take().unwrap_or("N/A");
```
Comment (NikVolf) [inline runner/src/runner.rs:210]: why not leave it `Option<&'static str>` instead of generating magic string?

File: runner/src/runner.rs [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+        for page in ext.memory_context.allocations().clone() {
+            let mut buf = vec![0u8; PageNumber::size()];
+            ext.get_mem(page.offset(), &mut buf);
+            let _ = program.set_page(page, &buf);
+        }
+
+        // Storing outgoing messages from message state.
+        let mut messages = Vec::new();
+
+        // We don't generate reply on trap, if we already processing trap message
```
Comment (NikVolf) [inline runner/src/runner.rs:250]: This thing can be based on `Outcome` and handled in upper layer

File: runner/src/runner.rs [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+    {
+        // Creating gas counter.
+        let mut gas_counter = GasCounter::new(message.gas_limit());
+
+        // Storing gas values.
+        let left_before = gas_counter.left();
+        let burned_before = gas_counter.burned();
+
+        // Charging for initial or loaded pages.
+        if settings.entry == EntryPoint::Init {
```
Comment (NikVolf) [inline runner/src/runner.rs:135]: I think entry point should coupled with `IncomingMessage`, not with settings

File: runner/src/runner.rs [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+
+use crate::configs::{AllocationsConfig, BlockInfo, EntryPoint};
+use crate::ext::Ext;
+use crate::ids::BlakeMessageIdGenerator;
+
+const EXIT_CODE_PANIC: i32 = 1;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ExecutionOutcome {
+    Done,
```
Comment (NikVolf) [inline runner/src/runner.rs:42]: where is `Wait` ?

File: runner/src/runner.rs (hunk 1 above)
Comment (breathx) [reply runner/src/runner.rs]: Option added back

File: runner/src/runner.rs (hunk 4 above)
Comment (breathx) [reply runner/src/runner.rs]: sticked in `Done` variant as a `wait` flag

File: core-processor/src/common.rs [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+            self.message_id(),
+            ERR_EXIT_CODE,
+        ))
+    }
+}
+
+#[derive(Debug)]
+pub enum JournalNote {
+    ExecutionFail {
+        origin: MessageId,
```
Comment (SabaunT) [inline core-processor/src/common.rs:157]: What do you think about changing field name? `origin` is related to substrate's account/message origin concept, so can lead to misunderstanding.

File: pallets/gear/src/manager.rs [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+}
+
+impl<T> ExtManager<T>
+where
+    T: Config,
+    T::AccountId: Origin,
+{
+    pub fn new() -> Self {
+        Default::default()
+    }
```
Comment (gshep) [inline pallets/gear/src/manager.rs:57]: stackoverflow

File: pallets/gear/src/manager.rs [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+        // Adjust block gas allowance
+        GasAllowance::<T>::mutate(|x| *x = x.saturating_sub(amount));
+
+        // TODO: weight to fee calculator might not be identity fee
+        let charge = T::GasConverter::gas_to_fee(amount);
+
+        self.gas_tree.as_mut().expect("Can't fail").spend(amount);
+
+        if let Err(e) = T::Currency::transfer(
+            &<T::AccountId as Origin>::from_origin(H256::from_slice(origin.as_slice())),
```
Comment (gshep) [inline pallets/gear/src/manager.rs:180]: Is it supposed to charge fee from a message?

Comment (gshep) [reply pallets/gear/src/manager.rs:176]: I guess the method should accept `origin: ProgramId` & `message_id: MessageId` so the message initiator could be charged

**Comment by NikVolf** [comment]
I think version bumping without migration is pointless, there is no way to upgrade runtime without storage migration, so we'll reset the network

**Review by NikVolf — APPROVED** [review]

**Review by shamilsan — APPROVED** [review]
Awesome work! 👍 

Just minor questions there.

File: core-processor/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+gear-core = { path = "../core" }
+gear-backend-common = { path = "../core-backend/common" }
+
+blake2-rfc = { version = "0.2.16", default-features = false }
+codec = { package = "parity-scale-codec", version = "2", features = ["derive"], default-features = false }
+log = { version = "0.4.14", default-features = false }
+
+[features]
+default = []
+strict = []
```
[the remark this answers is not in the document]
Comment (NikVolf) [reply core-processor/Cargo.toml]: `strict` one is needed to produce warnings on missing docs on the CI

**Pull request merged**